common/flatpak-appdata.c

```diff
@@ -187,6 +192,70 @@ start_element (GMarkupParseContext *context,
             }
         }
     }
+  else if (g_str_equal (element_name, "content_rating"))
+    {
+      /* https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-content_rating */
+      Component *component = NULL;
+
+      g_assert (data->components->len > 0);
+
+      component = g_ptr_array_index (data->components, data->components->len - 1);
+
+      if (component->content_rating == NULL)
+        {
+          const gchar *type = NULL;
+
+          if (g_markup_collect_attributes (element_name,
+                                                 attribute_names,
+                                                 attribute_values,
+                                                 error,
+                                                 G_MARKUP_COLLECT_STRING, "type", &type,
+                                                 G_MARKUP_COLLECT_INVALID))
+            {
+              component->content_rating_type = g_strdup (type);
+              component->content_rating = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, NULL);
+            }
+          else
+            {
+              g_warning ("Ignoring content rating missing type attribute");
+            }
+        }
+      else
+        {
+          g_warning ("Ignoring duplicate content rating");
+        }
+    }
+  else if (g_str_equal (element_name, "content_attribute"))
+    {
+      /* https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-content_rating */
+      Component *component = NULL;
+      const gchar *id = NULL;
+
```

Comment:
I don't like how this ignores the hierarchy of the file. We should set some in_content_rating boolean and check that in the child elements, otherwise you can create a file with e.g. a content_attribute outside a content_rating which will crash this code due to the hashtable not being allocated.

Reply:
Fixed locally.